### ISSUE TITLE
Compose Multiplatform support

### ISSUE THREAD

**Issue by tiwiz** [coil-kt/coil#842]
**Is your feature request related to a problem? Please describe.**
As an Android Developer, I would love to use Coil with Compose Desktop in the same way I use Coil for Compose on Android.

**Describe the solution you'd like**
Ideally, it would be the best to have the same approach we have on Android, so that the knowledge can be reused. 

**Additional context**
I think it would be OK to link the image download scope to either a `LaunchedEffect` in a `Composable`, or just link it to the Application lifecycle.

**Comment by colinrtwhite** [comment]
I agree this would be cool to have. That said, I think this is very very unlikely to be implemented as Coil relies on a number of Android framework classes for decoding (`BitmapFactory`, `ImageDecoder`) and integrates closely with `View`s and `Lifecycle`s. Also, Coil depends on a number of AndroidX libraries and OkHttp which aren't (and likely won't) be multiplatform. Overall, it's a huge amount of work!

Going to leave this open as a catch-all for multiplatform questions.

**Comment by jershell** [comment]
On `compose` I wrapped coil and wrote two `actual` functions. On desktop its look as

```kotlin
package components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.material.CircularProgressIndicator
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.DefaultAlpha
import androidx.compose.ui.graphics.ImageBitmap
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.layout.ContentScale
import com.tradernet.data.source.getHttpClient
import io.ktor.client.request.*
import org.jetbrains.skija.Image
import androidx.compose.foundation.Image as ComposeImage

suspend fun loadPicture(url: String): Result<ImageBitmap> {
    val client = getHttpClient()
    return try {
        val image = client.get<ByteArray>(url)
        Result.success(Image.makeFromEncoded(image).asImageBitmap())
    } catch (e: Exception) {
        Result.failure(e)
    }
}

@Composable
actual fun ExternalImage(url: String, modifier: Modifier, OnFail: @Composable () -> Unit) {
    var isLoading by remember { mutableStateOf(false) }
    var hasFail by remember { mutableStateOf(false) }
    var imageBitmap: ImageBitmap? by remember { mutableStateOf(null) }

    LaunchedEffect(url) {
        isLoading = true
        loadPicture(url)
            .onSuccess {
                imageBitmap = it
            }
            .onFailure {
                hasFail = true
            }
        isLoading = false
    }

    when {
        isLoading -> {
            Column(
                modifier = Modifier.fillMaxSize(),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.Center
            ) {
                CircularProgressIndicator()
            }
        }
        hasFail -> {
            OnFail()
        }
        else -> {
            imageBitmap?.let { bitmap ->
                
                ComposeImage(
                    bitmap = bitmap,
                    contentDescription = "",
                    contentScale = ContentScale.FillWidth,
                    alpha = DefaultAlpha,
                    colorFilter = null,
                )

            } ?: OnFail()
        }
    }
}
```

it's a variant I took from https://stackoverflow.com/questions/66002696/how-to-load-image-from-remote-url-in-kotlin-compose-desktop.
What's missing is caching mechanism.

**Comment by apkelly** [comment]
I too would love to see a coil-desktop implementation, even if it was just using JVM calls under the hood, similar to this post
https://dev.to/gerardpaligot/sharing-compose-components-between-android-and-desktop-17kg

**Comment by glureau** [comment]
Thanks @jershell . Just a note for those who will try to reuse your snippet: it was easier for me to use [skiko v0.5.3](https://github.com/JetBrains/skiko) instead of skija, and then just swap the line in the try catch with:

    Result.success(Image.makeFromEncoded(image).toComposeImageBitmap())

Works but yikes for the caching indeed, so +1 for Coil-desktop. Not sure if there is multiplatform alternatives or even work-in-progress on this topic, please share if you know something.

**Comment by yschimke** [comment]
In case it's relevant to the feasibility - https://github.com/square/okhttp/issues/6963

Libraries like Square Wire (gRPC) are also blocked on a multiplatform OkHttp Call.Factory. I hope in 2022, we get a good suite of KMP libraries for IO, Networking, RPC, Images, so hope this becomes feasible at some point.

**Comment by DevSrSouza** [comment]
For multiplatform projects there is the [Kamel](https://github.com/alialbaali/Kamel), based on watch I saw of the source code, almost all the source is based on `Ktor Client`, `Skiko` and `Androidx Compose`, this meas that even now that is working on Desktop and don't see how hard would be to make it support on iOS in the future.

**Comment by dragossusi** [comment]
Porting the library from `okhttp` to `ktor`(which can still use okhttp) will be a great start to make this library multiplatform.

Coroutines, ktor and okio are all multiplatform libraries.

**Comment by glureau** [comment]
Adding ktor means adding more code in Android apps with no purpose. Either Coil stick to Android and it should not try to move to ktor (and keep a lightweight performant library), or it's decided that Coil wants to go full KMP. I think a lot of Android projects will not move to KMP right now, so Coil is still a very good option as is.

Since Kamel is going multi-platform from start, maybe better to consider adding missing things to Kamel instead of trying to port a lot of Android specific code to KMP?

**Comment by evant** [comment]
This issue is for compose desktop which works on the jvm, not KMP, which I do think would be an easier ask.

**Comment by ursusursus** [comment]
Yes desktop (jvm) version would be great, all of infrastructure as far as coroutines, okhttp etc should be there, so "only" `Bitmap` & friends would need to be wrapped

Compose has a `ImageBitmap`

**Comment by GabrielLasso** [comment]
You can use libs like https://github.com/alialbaali/Kamel or https://github.com/ltttttttttttt/load-the-image and use @jershell's idea to wrap this libs with expect/actual.

**Comment by kazemcodes** [comment]
there is already an image loader that use coil as foundation to load images 
https://github.com/qdsfdhvh/compose-imageloader

**Comment by colinrtwhite** [comment]
Hi folks, big update! There are now plans to add [Kotlin Multiplatform](https://kotlinlang.org/docs/multiplatform.html) support to Coil and this will be the main feature of 3.x. I can't offer a timeline, but this is top of mind.

The plan is to leave Coil's public API largely the same (except for Android-specific classes) for an initial alpha then iterate on other public API changes we want to make for 3.x. Some of the main initial changes we'll need to make will be to abstract away usages of `Bitmap` and `Drawable` and replace `BitmapFactory` with a multiplatform solution. Fortunately, [Compose Multiplatform](https://github.com/JetBrains/compose-multiplatform) should make migrating the `coil-compose` artifact fairly straightforward. I'm heading on holiday until May 7th, but will follow up with more details afterwards.

**Comment by mxalbert1996** [comment]
To be clear, by Kotlin Multiplatform support do you mean support for Compose Multiplatform or some UI framework independent solution (which will be much more complicated I think)?
Also what platforms do you plan to support in addition to Android?

**Comment by eygraber** [comment]
@colinrtwhite have you considered applying for a grant for this - http://kotlinfoundation.org/grants

**Comment by colinrtwhite** [comment]
@mxalbert1996 The goal is to support Kotlin Multiplatform and Compose Multiplatform with as many targets as possible. The Android target will have the most functionality by supporting lifecycles, views, and other Android concepts. Other targets will be focused on supporting fetching + decoding images with integration with coroutines with less platform-specific support.

I think we can accomplish this by replacing `BitmapFactory` with [skiko](https://github.com/JetBrains/skiko), which is the image decoder used by Compose multiplatform, for non-Android targets. Migrating to Compose Multiplatform should be pretty straightforward once `coil-base` has been converted to multiplatform. We'll just need to swap AndroidX Compose with Jetbrains Compose.

The biggest challenge I see at the moment is refactoring `Context` references so `ImageLoader` and other classes aren't coupled with `Context`. Currently, we pass a `Context` object deep into `ImageLoader`'s classes.

**Comment by glureau** [comment]
Good news! Any idea if skiko will be more or less performant than BitmapFactory on Android?

For the Context I tend to use a ContentProvider that hold a static reference to the application Context, so that it can be used in androidMain code without being declared in commonMain. Do you see any issue with this approach? Is there a limitation with theme / image loading requiring the Activity/fragment context instead?

**Comment by MohamedRejeb** [comment]
@colinrtwhite Don't you think about making a development branch for working on this feature so we can contribute and make this process faster

**Comment by colinrtwhite** [comment]
@glureau My guess is it would be slightly slower given `BitmapFactory` is an Android framework class, which is why we should continue to use `BitmapFactory` on Android and skiko only on non-Android targets. For `Context` I'd prefer to avoid content providers and have the user pass a `Context` when building the `ImageLoader`.

@MohamedRejeb Will do! I'm going to create a `3.x` branch that folks can contribute to once I've handled some of the initial multiplatform migration.

**Comment by terrakok** [comment]
To use `BitmapFactory` on Android is the right decision because Skiko is not a lightweight library 😉

**Comment by sonatard** [comment]
@colinrtwhite Is there a plan tomigrate from OkHttp to Ktor for sending HTTP requests to support for Kotlin Multiplatform Mobile?

Or are you planning to use different HTTP clients depending on the environment?

**Comment by colinrtwhite** [comment]
@sonatard That's still up for discussion. I think it's likely we migrate to Ktor and make it an optional dependency to support https://github.com/coil-kt/coil/issues/1648.

**Comment by sonatard** [comment]
I'm relieved that Ktor has become the standard.

As many libraries are advancing their support for KMM using Ktor, it's pleasing to be able to reduce dependencies on other libraries like okHttp.

**Comment by natanfudge** [comment]
Why migrate to ktor if okhttp already supports Kotlin Multiplatform?

**Comment by ursusursus** [comment]
Hmm, is it? Stable?

**Comment by lucasmeneghin** [comment]
> I'm relieved that Ktor has become the standard.
> 
> As many libraries are advancing their support for KMM using Ktor, it's pleasing to be able to reduce dependencies on other libraries like okHttp.

+1 for using ktor

**Comment by mahramane** [comment]
pace yourself

**Comment by FunkyMuse** [comment]
I've been lurking this thread, just wanted to say congratulations https://kotlinfoundation.org/news/grants-program-winners-23/

**Comment by colinrtwhite** [comment]
@FunkyMuse Thanks! Huge thanks to the Kotlin team for selecting Coil for the grant. Also thanks to @eygraber for recommending I submit a proposal.

Quick update on multiplatform support: I've been hacking on support in the [`3.x`](https://github.com/coil-kt/coil/tree/3.x) branch and it's going fairly well. Core components like `DiskCache` and `MemoryCache` have already been migrated, but there's still plenty of work to do to integrate [Skiko](https://github.com/JetBrains/skiko) and refactor out references to Android classes. It looks like by and large the API should remain similar to `2.x`.

If folks are interested in helping out, I'd hold off for the moment until I can get it into a compilable state, though feel free to drop suggestions in this thread. I'll post again when it's ready.

**Comment by aaronbond** [comment]
great work dude :) really looking forward to it.

**Comment by DATL4G** [comment]
https://bugs.chromium.org/p/skia/issues/detail?id=12251

This probably needs to be addressed / fixed to support SVG on Desktop properly.

**Comment by yuroyami** [comment]
It's worth mentioning that Compose has a built-in platform-agnostic`ImageBitmap`, which is used as a parameter for the standard `Image` composable. Coil can make use of the open-source Compose code to see how this ImageBitmap is drawn.

**Comment by eygraber** [comment]
@colinrtwhite what's the status of the `3.x` branch? Is it compilable and/or usable? Have a couple of KMP side projects that are still in the POC stage that I'd like to use Coil with, and they don't need stability anytime soon (honestly don't even need it to really work as long as it compiles and doesn't crash).

**Comment by colinrtwhite** [comment]
@eygraber Hope to have it publishing snapshots soon-ish. The branch currently doesn't compile, but it's getting close. Tests are still mostly Android-only and will need to be updated. I'll start publishing snapshots as soon as it compiles + can load a basic network image. Will post here as soon as that's ready.

**Comment by colinrtwhite** [comment]
Hi folks, I just enabled snapshot publishing for Coil 3.0.0 to maven's snapshot repository:

```kotlin
allprojects {
    repositories {
        maven("https://oss.sonatype.org/content/repositories/snapshots")
    }
}

dependencies {
    implementation("io.coil-kt.coil3:coil:3.0.0-SNAPSHOT")
    implementation("io.coil-kt.coil3:coil-network:3.0.0-SNAPSHOT")
}
```

Documentation is still a WIP at the moment so here are some quick notes:

- Coil's network functionality was pulled into a separate module: `coil-network`. You'll need to manually add `NetworkFetcher.Factory()` to your image loader to load network images. Additionally, you'll need to add a [Ktor engine](https://ktor.io/docs/http-client-engines.html).
- `coil-base` was renamed to `coil-core` to align with the naming conventions used by Coroutines, Ktor, and AndroidX.
- All imports have been moved from `coil` -> `coil3`. This is to avoid binary compatibility issues and allow running Coil 2 and Coil 3 in the sample application side-by-side. That said, all APIs are a WIP and no binary compatibility guarantees are made before Coil's stable `3.0.0` release.
- Unlike Coil 2.x, `Cache-Control` headers aren't currently supported but will be eventually. There's also a new `CacheStrategy` interface where you can implement your own caching logic.
- WASM isn't supported yet (waiting on a couple dependencies), but it will be once those are updated.
- There are plenty of TODOs in the code and lots of opportunities for documentation improvements. Many of Coil's tests still only cover Android and could be converted to `commonMain`. Please feel free to submit PRs to the 3.x branch with bug fixes, documentation improvements, etc.!
- [Check out the samples](https://github.com/coil-kt/coil/tree/55709330a1a98af19ebdcb7f67b13bb092db7a81/samples) for examples of to use Coil with Compose Multiplatform and use configure your `ImageLoader`.

Will likely cut a proper `alpha01` in a few weeks after the snapshots are smoke tested for a few weeks and documentation is added.

**Comment by shibasis0801** [comment]
@colinrtwhite I wish to help with the WASM support. What are the pending tasks ?

**Comment by DRSchlaubi** [comment]
Probably ktor for once

**Comment by colinrtwhite** [comment]
@shibasis0801 Yep we're waiting on updates from Ktor, kotlinx-immutable-collections, Poko, and possibly others to support wasm.

**Comment by DRSchlaubi** [comment]
- https://youtrack.jetbrains.com/issue/KTOR-5587/Ktor-client-for-Kotlin-Wasm
- https://github.com/Kotlin/kotlinx.collections.immutable/issues/158
- https://github.com/drewhamilton/Poko/issues/235

Here a list of issues

**Comment by nowiwr01w** [comment]
@DRSchlaubi 
- Ktor client for K/Wasm works with `3.0.0-wasm2` 
- Looks like `kotlinx.collections.immutable` for K/Wasm was implemented [here](https://github.com/Kotlin/kotlinx.collections.immutable/pull/156)

So we're only waiting for [Poko](https://github.com/drewhamilton/Poko/issues/235)?

**Comment by FunkyMuse** [comment]
> Looks like `kotlinx.collections.immutable` for K/Wasm was implemented [here](https://github.com/Kotlin/kotlinx.collections.immutable/pull/156)

there still isn't a release of that PR

**Comment by colinrtwhite** [comment]
WIP `wasmJs` support is [here](https://github.com/coil-kt/coil/pull/1954), but runs into a compiler issue:

```
> Task :coil:wasmJsBrowserTest FAILED
Module not found: Error: Can't resolve './skiko.mjs'
```

If anyone knows how to work-around this feel free to add to that branch.

**Comment by inveders** [comment]
Hello @colinrtwhite maybe I made a mistake adding the snapshot but after adding the dependencies and the line for maven, I have the good coil3 package displaying, but inside there is no compose folder. So I can't reach AsyncImage and others things as in the sample. Is it normal?

**Comment by Kamil-H** [comment]
@inveders 
I think I faced the same problem. I had to change the JetBrains Compose version to `1.6.0-alpha01`, and it worked. In fact, I faced another issue:
```
Could not find org.jetbrains.skiko:skiko:0.7.90
```
but this one, I fixed by strictly declaring `skiko` version:
```
commonMainImplementation("org.jetbrains.skiko:skiko") {
    version {
        strictly("0.7.89.1")
    }
}
```

**Comment by colinrtwhite** [comment]
@inveders You need to add a dependency on Coil's Compose artifact:

```kotlin
implementation("io.coil-kt.coil3:coil-compose:3.0.0-SNAPSHOT")
```

**Comment by rustamsmax** [comment]
@colinrtwhite 
Is there a plan to support svg artifact for multiplatform (android/ios)?
I'd like to test coil3 in my projects but loading svg images is very crucial for it

**Comment by colinrtwhite** [comment]
@rustamsmax No plans yet as I'm focused improving on Coil's core and Compose integration.

I think it's possible to make `coil-svg` multiplatform using `org.jetbrains.skia.svg.SVGDOM` on non-Android platforms. I'd be open to a PR that adds support for this!

 `coil-gif` and `coil-video` are unlikely to be multiplatform as I don't think Skiko provides animated image support or video frame support.

**Comment by DRSchlaubi** [comment]
> @rustamsmax No plans yet as I'm focused improving on Coil's core and Compose integration.
> 
> I think it's possible to make `coil-svg` multiplatform using `org.jetbrains.skia.svg.SVGDOM` on non-Android platforms. I'd be open to a PR that adds support for this!
> 
> `coil-gif` and `coil-video` are unlikely to be multiplatform as I don't think Skiko provides animated image support or video frame support.

There is GIF support in [this awsome lib](https://github.com/qdsfdhvh/compose-imageloader/blob/master/image-loader/src/skiaMain/kotlin/com/seiko/imageloader/component/decoder/GifDecoder.kt) so there seems to be a way

**Comment by rustamsmax** [comment]
@colinrtwhite org.jetbrains.skia.svg.SVGDOM has "render" method which draws image to the canvas.
But the new Image class(for non android targets) return Bitmap. So it's a bit difficult to write a decoder for svg

**Comment by colinrtwhite** [comment]
@rustamsmax Ah yep, I think the `Image` API is likely to change after the first alpha is released. Currently it's only built to support static images on non-Android platforms. For the moment, you could prerender the SVG to a bitmap since we know the output size (`options.size`) and return that from the `Decoder`.

**Comment by MohamedRejeb** [comment]
I'm working on migrating `coil-svg` to multiplatform and it's almost done. You can follow the progress here: #1965

**Comment by MohamedRejeb** [comment]
I'm thinking of changing the `Image` API by an interface similar to Compose `Painter`, it can be called `CoilPainter` or just `Painter` and it has a size and an `onDraw` method.
And in `coil-compose-core` this coil `Painter` is going to behave the same as Compose `Painter`, it's just a wrapper after all.
With this change, we will be able to migrate `coil-gif` to multiplatform, and `coil-svg` in multiplatform will be more lightweight because we will draw the svg to the canvas directly, no need to create Bitmap and Image instances.

**Comment by rustamsmax** [comment]
@MohamedRejeb maybe best approach is to make it a sealed class? any time in the future we can add extra type

**Comment by MohamedRejeb** [comment]
Yes, it's possible, but for Compose `Painter` should cover all the cases. This change is only for non-Android targets which are all using Compose Multiplatform.

**Comment by alekseyHunter** [comment]
@colinrtwhite
Hello!
Can you please tell my, if there are any plans to support displaying local images?

**Comment by colinrtwhite** [comment]
Hi folks, I just published Coil `3.0.0-alpha01`. I'd recommend using that version over the snapshots. Check out the WIP upgrade guide [here](https://coil-kt.github.io/coil/upgrading_to_coil3/).

I haven't merged wasm support yet as I'm still waiting for Ktor support. If you need wasm support I'd continue to depend on the snapshots, which use Ktor `3.0.0-wasm2`.

**Comment by nowiwr01w** [comment]
@colinrtwhite it would be great if you'd publish a snapshot with `3.0.0-wasm02`

**Comment by rustamsmax** [comment]
@colinrtwhite https://coil-kt.github.io/coil/upgrading_to_coil3 link is not working

<img width="604" alt="image" src="https://github.com/coil-kt/coil/assets/36570577/46fe32d7-a6a2-42e9-9ec9-728ec22db6be">

**Comment by DRSchlaubi** [comment]
It's [here](https://github.com/coil-kt/coil/blob/3.x/docs/upgrading_to_coil3.md)

**Comment by sonatard** [comment]
Hi! We're having trouble finding coil3.compose. Are there any other changes needed to migrate to Coil 3?

```
e: file:///xxx/app/src/main/java/com/xxx/android/components/module/Image.kt:6:14 Unresolved reference: compose
```

```diff
diff --git a/app/build.gradle.kts b/app/build.gradle.kts
index c0947d79f..9e013c1d4 100644
--- a/app/build.gradle.kts
+++ b/app/build.gradle.kts
@@ -1018,7 +1018,9 @@ dependencies {
-    implementation("io.coil-kt:coil-compose:2.3.0")
+    val coilVersion = "3.0.0-alpha01"
+    implementation("io.coil-kt.coil3:coil:$coilVersion")
+    implementation("io.coil-kt.coil3:coil-network:$coilVersion")
 
diff --git a/app/src/main/java/com/xxx/android/components/module/Image.kt b/app/src/main/java/com/xxx/android/components/module/Image.kt
index 3dde784eb..e2e45ca94 100644
--- a/app/src/main/java/com/xxx/android/components/module/Image.kt
+++ b/app/src/main/java/com/xxx/android/components/module/Image.kt
@@ -3,9 +3,9 @@ package com.xxx.android.components.module
-import coil.compose.AsyncImagePainter
-import coil.compose.SubcomposeAsyncImage
-import coil.compose.SubcomposeAsyncImageContent
+import coil3.compose.AsyncImagePainter
+import coil3.compose.SubcomposeAsyncImage
+import coil3.compose.SubcomposeAsyncImageContent
 
 @Composable
 fun Image(
```

**Comment by colinrtwhite** [comment]
@sonatard You need to add a dependency on Coil's Compose artifact:

```kotlin
implementation("io.coil-kt.coil3:coil-compose:3.0.0-alpha01")
```

**Comment by sonatard** [comment]
@colinrtwhite Thanks!!  I successfully built it by adding a dependency.

**Comment by sonatard** [comment]
@colinrtwhite Thank you for your support with KMP!

Coil3 is now running, so here's my diff.  I hope you find it useful.

```diff
diff --git a/app/build.gradle.kts b/app/build.gradle.kts
--- a/app/build.gradle.kts
+++ b/app/build.gradle.kts
@@ -1018,7 +1018,9 @@ dependencies {

-    implementation("io.coil-kt:coil-compose:2.3.0")
+    val coilVersion = "3.0.0-alpha01"
+    implementation("io.coil-kt.coil3:coil-compose:$coilVersion")
+    implementation("io.coil-kt.coil3:coil-network:$coilVersion")
 
diff --git a/app/src/main/java/com/xxx/android/components/Main.kt b/app/src/main/java/com/xxx/android/components/Main.kt
index 1bfe7a053..0aab82fa6 100644
--- a/app/src/main/java/com/xxx/android/components/Main.kt
+++ b/app/src/main/java/com/xxx/android/components/Main.kt
@@ -1,11 +1,25 @@
+import coil3.ImageLoader
+import coil3.annotation.ExperimentalCoilApi
+import coil3.compose.setSingletonImageLoaderFactory
+import coil3.fetch.NetworkFetcher

 
+@OptIn(ExperimentalCoilApi::class)
 @Composable
 fun Main(infra: Infra, initialUrl: String?) {
+    // Coil3
+    // https://coil-kt.github.io/coil/upgrading_to_coil3/#network-images
+    setSingletonImageLoaderFactory { context ->
+        ImageLoader.Builder(context)
+            .components {
+                add(NetworkFetcher.Factory())
+            }
+            .build()
+    }
diff --git a/app/src/main/java/com/xxx/android/components/module/Image.kt b/app/src/main/java/com/xxx/android/components/module/Image.kt
--- a/app/src/main/java/com/xxx/android/components/module/Image.kt
+++ b/app/src/main/java/com/xxx/android/components/module/Image.kt
-import coil.compose.AsyncImagePainter
-import coil.compose.SubcomposeAsyncImage
-import coil.compose.SubcomposeAsyncImageContent
+import coil3.compose.AsyncImagePainter
+import coil3.compose.SubcomposeAsyncImage
+import coil3.compose.SubcomposeAsyncImageContent
 
 @Composable
 fun Image(

**Comment by Matcha-xiaobin** [comment]
你好，在ios上提示kotlin.IllegalStateException: TLS sessions are not supported on Native platform.
这个要怎么解决？

**Comment by Vaibhav2002** [comment]
VectorPainters when used in placeholder, does not pick the LocalContentColor and is always in black color
Hence not being visible in dark mode

```
placeholder = rememberVectorPainter(Icons.Default.Domain),
```

**Comment by colinrtwhite** [comment]
@Vaibhav2002 Please file a separate issue for this

**Comment by mainrs** [comment]
Regarding gif support, there is an official UI elements from Jetbrains that can render GIFs: https://github.com/JetBrains/compose-multiplatform/tree/master/components/AnimatedImage

**Comment by mahramane** [comment]
When will Wasm be supported?

**Comment by colinrtwhite** [comment]
@mahramane I'm waiting for Ktor to publish a version with wasm support to Maven Central. I'm hesitant to publish a separate version to Maven Central like `3.0.0-alpha04-wasm` as it's possible Renovate and other automated update services will try to update to those instead of the standard alphas.

If you need wasm support today the snapshots support WASM via `3.0.0-SNAPSHOT`.

**Comment by 19lmyers** [comment]
> @mahramane I'm waiting for Ktor to publish a version with wasm support to Maven Central. I'm hesitant to publish a separate version to Maven Central like `3.0.0-alpha04-wasm` as it's possible Renovate and other automated update services will try to update to those instead of the standard alphas.
> 
> If you need wasm support today the snapshots support WASM via `3.0.0-SNAPSHOT`.

Hello, I'm attempting to pull in the snapshot version for WASM, but for some reason Gradle gives me an error claiming that the snapshot `io.coil-kt.coil3:coil-network-ktor-wasm-js:3.0.0-SNAPSHOT:20240215.162906-49` is using Ktor 2.3.8, which doesn't support WASM.

**Comment by artemyto** [comment]
Wasm support has been merged to Ktor. The next 3.0.0 beta release of Ktor should support Wasm.

**Comment by AmeerAmjed** [comment]
Any update???
 I use **Ktor** version **3.0.0-wasm2** but can't install coil3, I'm try install mutli version but get error "3.0.0-alpha04-wasm" &  "3.0.0-alpha04" which one support wasmjs.

**Comment by colinrtwhite** [comment]
@AmeerAmjed As mentioned [here](https://github.com/coil-kt/coil/issues/842#issuecomment-1922332425) that version doesn't exist

**Comment by colinrtwhite** [comment]
@AmeerAmjed `3.0.0-alpha04-wasm` isn't a [valid version](https://github.com/coil-kt/coil/issues/842#issuecomment-1922332425).

**Comment by colinrtwhite** [comment]
`3.0.0-alpha05` adds support for `wasmJs`! I figured out a [work-around](https://github.com/coil-kt/coil/blob/7c42d05978ef38433bfc72e40f3a98d5b08bd8ab/buildSrc/src/main/kotlin/coil3/multiplatform.kt#L128) so Ktor `3.0.0-wasm2` (which isn't available on `mavenCentral`) is only required for the `wasmJs` artifact. If you don't add the `wasmJs` target, Coil will depend on Ktor `2.3.8`.

**Comment by joreilly** [comment]
@colinrtwhite I probably misunderstood something from above but getting following when depending on `3.0.0-alpha05`


```Could not resolve all dependencies for wasmJsCompileClasspath.
> Could not find io.coil-kt:coil:3.0.0-alpha05.
  Searched in the following locations:
```

**Comment by joreilly** [comment]
ah, I did have wrong dependency....following works

```
implementation("io.coil-kt.coil3:coil-compose:3.0.0-alpha05")
```

**Comment by joreilly** [comment]
So, it's resolving but image isn't being displayed for some reason.  I'll try to dig a bit deeper later but if you do want to try it's happening in Compose for Web client in following branch 

https://github.com/joreilly/PeopleInSpace/tree/dependency_updates

**Comment by rustamsmax** [comment]
@joreilly maybe we must pass string as Uri

**Comment by romainbsl** [comment]
@joreilly I had the same issue when migrating to coil3.

Network feature has been extracted from `coil-core` so you need to had a network dependency
https://coil-kt.github.io/coil/upgrading_to_coil3/#network-images

Surely you need `io.coil-kt.coil3:coil-network-ktor` :D

**Comment by joreilly** [comment]
Thanks @romainbsl , it's working now!

**Comment by sunny7498** [comment]
Getting this error in compose for desktop
<img width="747" alt="image" src="https://github.com/coil-kt/coil/assets/99462378/ddd334ab-a1fa-46f1-a636-c778dd574b1b">

**Comment by MohamedRejeb** [comment]
You need to add kotlinx-coroutines-android to your androidMain and kotlinx-coroutines-swing to your desktopMain @sunny7498

**Comment by hakanai** [comment]
Been trying for a while today to make a better replacement for this error dialog. Good grief... it's like JetBrains have never heard of SwingX.

**Comment by hakanai** [comment]
Yeah I'm not saying to haul in the whole library for an error dialog, but to imitate it with one written in Compose which looks exactly like it

**Comment by RecodeLiner** [comment]
Good afternoon everyone, I am translating a multiplatform project from compose image loader to kamel, and from it I decided to translate it to coil, I took the implementation on ktor and ran into a problem:
Locally in IDE everything works fine, release on android also gives images correctly, but when I make release on desktop(msi for windows) I face the problem that the same link that works well for me in ide and in browser, for some reason gives error condition

**Comment by DRSchlaubi** [comment]
Sounds like proguard maybe?

**Comment by RecodeLiner** [comment]
Yes, I thought about that too, but to be honest, I don’t even know what could cause this, because coil documentation only says about coroutines and okhttp, but they are already specified in my proguard rules file

**Comment by DRSchlaubi** [comment]
It would be useful to actually see the error,

Coil components use java services, so maybe proguard erases them

**Comment by lepicekmichal** [comment]
Hi, trying to use it on iOS (alpha6) and all is good but one thing, svg. While they do load, they are all tinted black.
Am I doing something wrong? How do you install the decoder? I've got this:

```
setSingletonImageLoaderFactory {
    ImageLoader.Builder(it)
        .components { add(SvgDecoder.Factory()) }
        .build()
}
```

I'll create a separate bug if this is all it should take, thank you!

**Comment by DatL4g** [comment]
@lepicekmichal 
Known skia issue https://issues.skia.org/issues/40043341

**Comment by master-lzh** [comment]
Does anyone know how to use `AsyncImage` on iOS platform and apply rounded corners or circles through `transformations` like in android

**Comment by colinrtwhite** [comment]
@master-lzh You should use `Modifier.clip` [for this](https://github.com/coil-kt/coil/blob/4426b434e42ea6d8c910113535ff4c228893d444/coil-core/src/androidMain/kotlin/coil3/transform/RoundedCornersTransformation.kt#L28).

**Comment by master-lzh** [comment]
> @master-lzh You should use `Modifier.clip` [for this](https://github.com/coil-kt/coil/blob/4426b434e42ea6d8c910113535ff4c228893d444/coil-core/src/androidMain/kotlin/coil3/transform/RoundedCornersTransformation.kt#L28). Transformations overall are obsolescenced by Compose's `Modifier.graphicsLayer` API.

Thanks a lot, I tried it yesterday and it's working now

**Comment by nickfaces** [comment]
I tested 3.0.0-alpha06 on android and  iOS, but I can't understand how to set the headers for setting auth cookies.

**Comment by kasem-sm** [comment]
Anyone experiencing high CPU usage on iOS with setSingletonImageLoaderFactory?

**Comment by guillermolc** [comment]
Hi everyone i received this error on my wasm web impl 
`Access to fetch at ... from origin 'http://localhost:8080' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.`

**Comment by joreilly** [comment]
Some related info here fwiw https://github.com/qdsfdhvh/compose-imageloader/issues/446

**Comment by guillermolc** [comment]
Hi @joreilly i checked that but isn't working
The image loader class have a constructor that require a context and it doesn't have a lambda to add new components only can read those

**Comment by realityexpander** [comment]
> Hi everyone i received this error on my wasm web impl `Access to fetch at ... from origin 'http://localhost:8080' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.`
> 
> im using 3.0.0-alpha06 version on my project, any tip to solve this?

Did you set up CORS on your server to allow the request to go thru on the browser side?

**Comment by mahramane** [comment]
[3.0.0-beta-2-eap-932](https://mvnrepository.com/artifact/io.ktor/ktor-client-core-wasm-js/3.0.0-beta-2-eap-932) is out and support wasm js.
please update the ktor version in coil library.

**Comment by artemyto** [comment]
> [3.0.0-beta-2-eap-932](https://mvnrepository.com/artifact/io.ktor/ktor-client-core-wasm-js/3.0.0-beta-2-eap-932) is out and support wasm js.
> please update the ktor version in coil library.

This version is not on Maven Central, but is from different repository (KtorEAP)

**Comment by sivansundar** [comment]
For some reason Im not able to load images despite using the right versions and dependencies. 

Using the following under my `commonMain` : 

```
implementation("io.coil-kt.coil3:coil-compose:3.0.0-alpha06")
implementation("io.coil-kt.coil3:coil-network-ktor:3.0.0-alpha06")
implementation("io.ktor:ktor-client-core:3.0.0-wasm2")

```

Composable : 

```
 AsyncImage(
        modifier = Modifier.wrapContentSize(),
        model = item.imageUrl,
        contentDescription = "",
        error = painterResource(resource = Res.drawable.compose_multiplatform)
)
```

Not sure if I am missing something very basic.

**Comment by sivansundar** [comment]
Sorry about that.

```
🚨 Failed - https://storage.googleapis.com/input-promo-banners/flyer-templates/syngenta_0529_a.webp - java.lang.ExceptionInInitializerError
2024-05-29 21:28:31.009  7146-7146  RealImageLoader         dev.sivan.muncheese                  I  🚨 Failed - <IMAGEURL> - java.lang.NoClassDefFoundError: <clinit> failed for class io.ktor.client.HttpClientJvmKt; see exception in other thread
2024-05-29 21:28:31.009  7146-7146  RealImageLoader         dev.sivan.muncheese                  I  🚨 Failed - <IMAGEURL> - java.lang.NoClassDefFoundError: io.ktor.client.HttpClientJvmKt
2024-05-29 21:28:31.009  7146-7146  RealImageLoader         dev.sivan.muncheese                  I  🚨 Failed - <IMAGEURL> - java.lang.NoClassDefFoundError: io.ktor.client.HttpClientJvmKt
```

**Comment by DRSchlaubi** [comment]
> Sorry about that.
> 
> ```
> 🚨 Failed - <IMAGEURL> - java.lang.ExceptionInInitializerError
> 2024-05-29 21:28:31.009  7146-7146  RealImageLoader         dev.sivan.muncheese                  I  🚨 Failed - <IMAGEURL> - java.lang.NoClassDefFoundError: <clinit> failed for class io.ktor.client.HttpClientJvmKt; see exception in other thread
> 2024-05-29 21:28:31.009  7146-7146  RealImageLoader         dev.sivan.muncheese                  I  🚨 Failed - <IMAGEURL> - java.lang.NoClassDefFoundError: io.ktor.client.HttpClientJvmKt
> 2024-05-29 21:28:31.009  7146-7146  RealImageLoader         dev.sivan.muncheese                  I  🚨 Failed - <IMAGEURL> - java.lang.NoClassDefFoundError: io.ktor.client.HttpClientJvmKt
> ```

Are you using proguard by chance?

**Comment by sivansundar** [comment]
No.

**Comment by sivansundar** [comment]
Alright I made it work. Missed adding ktor clients to all the platforms. Now my dependencies looks like this : 

```
        commonMain.dependencies {
            implementation("io.coil-kt.coil3:coil:3.0.0-alpha06")
            implementation("io.coil-kt.coil3:coil-compose:3.0.0-alpha06")
            implementation("io.coil-kt.coil3:coil-network-ktor:3.0.0-alpha06")

            implementation("io.ktor:ktor-client-core:3.0.0-wasm2")
        }

        androidMain.dependencies {
            implementation("io.coil-kt.coil3:coil-network-okhttp:3.0.0-alpha06")

        }

        desktopMain.dependencies {
            implementation(compose.desktop.currentOs)
            implementation("org.jetbrains.kotlinx:kotlinx-coroutines-swing")
            implementation("io.coil-kt.coil3:coil-network-okhttp:3.0.0-alpha06")

        }

        iosMain.dependencies {
            implementation("io.ktor:ktor-client-darwin:3.0.0-wasm2")
        }

        wasmJsMain.dependencies {
            implementation("io.ktor:ktor-client-js:3.0.0-wasm2")
        }
```

**Comment by Feinate** [comment]
How to use png file from composeResources with coil... Seems to be impossible, lack of documentation...

**Comment by ColtonIdle** [comment]
One liner for compose multiplatform https://github.com/alexstyl/addcoilmultiplatform

**Comment by skydoves** [comment]
[Landscapist](https://github.com/skydoves/landscapist?tab=readme-ov-file#coil) also supports KMP with coil3.

```kotlin
sourceSets {
    val commonMain by getting {
        dependencies {
            implementation("com.github.skydoves:landscapist-coil3:2.3.6")
        }
    }
}
```

**Comment by joreilly** [comment]
Hopefully this is ok place to post this but has anyone come across error like this when using Coil3 in a Wasm based Compose for Web client?  Seeing this currently in the PeopleInSpace sample.

```
illegal cast
RuntimeError: illegal cast
    at <com.example:compose-web>.coil3.network.ktor.internal.KtorNetworkResponseBody__close-impl (http://localhost:8080/peopleinspace.wasm:wasm-function[67145]:0x8f8ac0)
    at <com.example:compose-web>.coil3.network.ktor.internal.KtorNetworkResponseBody.close (http://localhost:8080/peopleinspace.wasm:wasm-function[67151]:0x8f8bee)
    at <com.example:compose-web>.kotlin.closeFinally (http://localhost:8080/peopleinspace.wasm:wasm-function[7234]:0x413c78)
    at <com.example:compose-web>.coil3.network.internal.$readBufferCOROUTINE$3.doResume (http://localhost:8080/peopleinspace.wasm:wasm-function[63020]:0x8a71ef)
    at <com.example:compose-web>.coil3.network.internal.readBuffer (http://localhost:8080/peopleinspace.wasm:wasm-function[63015]:0x8a6ddd)
    at <com.example:compose-web>.coil3.network.NetworkFetcher$fetch$slambda.doResume (http://localhost:8080/peopleinspace.wasm:wasm-function[62948]:0x8a4d0d)
    at <com.example:compose-web>.coil3.network.NetworkFetcher$fetch$slambda.invoke (http://localhost:8080/peopleinspace.wasm:wasm-function[62946]:0x8a4aae)
    at <com.example:compose-web>.coil3.network.NetworkFetcher$fetch$slambda.invoke (http://localhost:8080/peopleinspace.wasm:wasm-function[62947]:0x8a4adc)
    at <com.example:compose-web>.coil3.network.NetworkFetcher$executeNetworkRequest$slambda.doResume (http://localhost:8080/peopleinspace.wasm:wasm-function[62960]:0x8a5244)
    at <com.example:compose-web>.coil3.network.NetworkFetcher$executeNetworkRequest$slambda.invoke (http://localhost:8080/peopleinspace.wasm:wasm-function[62958]:0x8a5138)
```

**Comment by colinrtwhite** [comment]
@joreilly Can you try with the [latest snapshot](https://coil-kt.github.io/coil/faq/#how-do-i-get-development-snapshots)? I just bumped main to Ktor `3.0.0-rc-1`, which might fix it.

**Comment by joreilly** [comment]
@colinrtwhite ah, my bad, I messed up when updating to alpha10.  Apologies for the confusion.

**Comment by JacobCube** [comment]
Coil version 3.0.0-alpha10 together with Ktor version 3.0.0-rc-1 is crashing for me on release Jvm. See printstack below.
`java.lang.ClassFormatError: Invalid index 2 in LocalVariableTable in class file coil3/compose/internal/UtilsKt
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1017)
	at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:150)
	at java.base/jdk.internal.loader.BuiltinClassLoader.defineClass(BuiltinClassLoader.java:862)
	at java.base/jdk.internal.loader.BuiltinClassLoader.findClassOnClassPathOrNull(BuiltinClassLoader.java:760)
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClassOrNull(BuiltinClassLoader.java:681)
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:639)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
	at coil3.compose.AsyncImageKt.AsyncImage-76YX9Dk(AsyncImage.kt:161)
	at coil3.compose.AsyncImageKt.AsyncImage-QgsmV_s(AsyncImage.kt:132)
	at coil3.compose.SingletonAsyncImageKt.AsyncImage-gl8XCv8(SingletonAsyncImage.kt:125)`


Even though my Proguard rules contain both `-keep class coil3.** { *; }` and `-keep class io.ktor.client.** { *; }
-keep class io.ktor.utils.io.** { *; }
-keep class io.ktor.utils.io.jvm.** { *; }
-keep class io.ktor.utils.io.nio.** { *; }
-keep class io.ktor.server.config.** { *; }
-keep class io.ktor.serialization.** { *; }`

**Comment by colinrtwhite** [comment]
@JacobCube I would avoid adding `-keep class coil3.** { *; }` as that prevents all optimization, obfuscation, and tree shaking of Coil's code. Compose code specifically really benefits from R8's optimization. For your error I'd recommend reporting it to the R8 issue tracker. You could also try using a different version of AGP as it usually bundles newer versions of R8 with fixes.

**Comment by JacobCube** [comment]
Thanks for the feedback, I'll change the rules. Regarding the R8, the issue occurs only with this library, so it's either some missing rule I thought you'd know of or some internal issue. Am I missing something? I'll report it to R8 nonetheless.

**Comment by StefanOltmann** [comment]
I don't understand how Coil3 is supposed to be used for SVGs.
My goal is to display a SVG on Desktop & WASM.

I added these dependencies to `commonMain`:
```
implementation("io.coil-kt.coil3:coil-compose:3.0.0-alpha10")
implementation("io.coil-kt.coil3:coil-svg:3.0.0-alpha10")
```

I set the SVG factory in my App() composable:
```
setSingletonImageLoaderFactory {
    ImageLoader.Builder(PlatformContext.INSTANCE)
        .components {
            add(SvgDecoder.Factory())
        }
        .build()
}
```

But this displays always a black screen:
```
AsyncImage(
    model = mySvgString, // "<svg... and so on>"
    contentDescription = null
)
```

What's the correct way to display SVGs using Coil?

**Comment by colinrtwhite** [comment]
Hi folks, I'm going to close this issue now that Compose Multiplatform is fully supported in version `3.0.0-rc01`. If you find bugs please open any new issues [here](https://github.com/coil-kt/coil/issues/new/choose). Thanks!
